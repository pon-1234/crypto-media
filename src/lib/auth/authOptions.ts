import { NextAuthOptions } from 'next-auth'
import GoogleProvider from 'next-auth/providers/google'
import EmailProvider from 'next-auth/providers/email'
import { FirestoreAdapter } from '@next-auth/firebase-adapter'
import { cert } from 'firebase-admin/app'
import { isTestOrCI } from '@/lib/env/detect'

/**
 * Firebase設定が有効かどうかを判定
 */
const hasValidFirebaseConfig = 
  process.env.FIREBASE_ADMIN_PRIVATE_KEY && 
  process.env.FIREBASE_ADMIN_PRIVATE_KEY !== 'test-private-key'

/**
 * NextAuth.js設定
 * @doc https://next-auth.js.org/configuration/options
 * @related src/app/api/auth/[...nextauth]/route.ts - NextAuth APIルート
 * @issue #1 - 初期セットアップ：NextAuth.jsによる認証実装
 */
export const authOptions: NextAuthOptions = {
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    }),
    EmailProvider({
      server: {
        host: process.env.EMAIL_SERVER_HOST,
        port: Number(process.env.EMAIL_SERVER_PORT),
        auth: {
          user: process.env.EMAIL_SERVER_USER,
          pass: process.env.EMAIL_SERVER_PASSWORD,
        },
      },
      from: process.env.EMAIL_FROM,
    }),
  ],
  // テスト環境またはFirebase未設定の場合はFirestoreAdapterを使用しない
  ...(!isTestOrCI() && hasValidFirebaseConfig && {
    adapter: FirestoreAdapter({
      credential: cert({
        projectId: process.env.FIREBASE_ADMIN_PROJECT_ID!,
        clientEmail: process.env.FIREBASE_ADMIN_CLIENT_EMAIL!,
        privateKey: process.env.FIREBASE_ADMIN_PRIVATE_KEY!.replace(
          /\\n/g,
          '\n'
        ),
      }),
    }),
  }),
  session: {
    strategy: 'jwt',
  },
  callbacks: {
    async session({ session, token }) {
      if (session?.user && token.sub) {
        session.user.id = token.sub
      }
      return session
    },
    async jwt({ user, token }) {
      if (user) {
        token.uid = user.id
      }
      return token
    },
  },
  pages: {
    signIn: '/login',
    signOut: '/login',
    error: '/login',
    verifyRequest: '/login?verify=1',
  },
}
