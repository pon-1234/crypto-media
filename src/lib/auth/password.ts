import bcryptjs from 'bcryptjs'

/**
 * パスワードのハッシュ化と検証に関するユーティリティ
 * @doc DEVELOPMENT_GUIDE.md#認証・会員DB - パスワード認証のセキュリティ実装
 * @issue #26 - 認証機能の拡張：メール/パスワード認証の追加
 */

/**
 * パスワードをハッシュ化する
 * @param password プレーンテキストのパスワード
 * @returns ハッシュ化されたパスワード
 * @security-note 本番環境では必ずsaltRounds=10を使用してセキュリティを確保
 */
export async function hashPassword(password: string): Promise<string> {
  // テスト環境では処理を高速化するためsaltRoundsを減らす
  // 本番環境のセキュリティには影響しない
  // @security-note: 本番環境では必ずsaltRounds=10を使用
  const saltRounds = process.env.NODE_ENV === 'test' ? 1 : 10
  return bcryptjs.hash(password, saltRounds)
}

/**
 * パスワードを検証する
 * @param password プレーンテキストのパスワード
 * @param hashedPassword ハッシュ化されたパスワード
 * @returns パスワードが一致する場合はtrue
 */
export async function verifyPassword(
  password: string,
  hashedPassword: string
): Promise<boolean> {
  return bcryptjs.compare(password, hashedPassword)
}

/**
 * パスワードの強度を検証する
 * @param password パスワード
 * @returns 検証結果
 */
export function validatePasswordStrength(password: string): {
  isValid: boolean
  errors: string[]
} {
  const errors: string[] = []

  if (password.length < 8) {
    errors.push('パスワードは8文字以上で入力してください')
  }

  if (!/[A-Z]/.test(password)) {
    errors.push('大文字を1文字以上含めてください')
  }

  if (!/[a-z]/.test(password)) {
    errors.push('小文字を1文字以上含めてください')
  }

  if (!/[0-9]/.test(password)) {
    errors.push('数字を1文字以上含めてください')
  }

  if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
    errors.push('特殊文字を1文字以上含めてください')
  }

  return {
    isValid: errors.length === 0,
    errors,
  }
}
