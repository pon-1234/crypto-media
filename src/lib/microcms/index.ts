/**
 * microCMS ライブラリのエントリーポイント
 * @doc DEVELOPMENT_GUIDE.md#microCMS
 * @issue #2 - microCMSクライアントと型定義の実装
 */

// クライアント
export { client, defaultQueries, MAX_LIMIT, getOptimizedImageUrl } from './client'

// スキーマとバリデーション
export {
  // スキーマ
  corporateNewsSchema,
  expertSchema,
  categorySchema,
  tagSchema,
  featureSchema,
  mediaArticleSchema,
  siteSettingsSchema,
  researchSurveySchema,
  pageCorporateSchema,
  
  // リストレスポンススキーマ
  corporateNewsListSchema,
  expertListSchema,
  categoryListSchema,
  tagListSchema,
  featureListSchema,
  mediaArticleListSchema,
  researchSurveyListSchema,
  pageCorporateListSchema,
  
  // ヘルパー関数
  createListResponseSchema,
  
  // 型
  type CorporateNews,
  type Expert,
  type Category,
  type Tag,
  type Feature,
  type MediaArticle,
  type SiteSettings,
  type ResearchSurvey,
  type PageCorporate,
} from '@/lib/schema'

// 追加の型エクスポート
export type {
  MicroCMSBase,
  MicroCMSListResponse,
  MicroCMSImage,
  ExpertRole,
  ArticleType,
  MembershipLevel,
} from '@/lib/schema'

// APIメソッド
export {
  getCorporateNewsList,
  getCorporateNewsDetail,
  getAllCorporateNewsIds,
} from './corporate-news'