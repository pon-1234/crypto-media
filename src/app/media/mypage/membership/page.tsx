import { Metadata } from 'next'
import { getServerSession } from 'next-auth'
import { authOptions } from '@/lib/auth/authOptions'
import { getUserMembership } from '@/lib/auth/membership'
import { format } from 'date-fns'
import { ja } from 'date-fns/locale'
import Link from 'next/link'
import { Button } from '@/components/ui/button'
import {
  Crown,
  ArrowLeft,
  CreditCard,
  Calendar,
  AlertCircle,
  CheckCircle,
  XCircle,
} from 'lucide-react'

export const metadata: Metadata = {
  title: '会員情報・お支払い | Crypto Media',
  description: '会員プランの確認・変更、お支払い情報の管理',
}

/**
 * 会員管理ページ
 *
 * @doc 会員プランの確認・変更、Stripeカスタマーポータルへのアクセスを提供
 * @related src/app/api/stripe/portal/route.ts - Stripeポータルへのリダイレクト
 * @related src/lib/auth/membership.ts - 会員情報取得
 * @issue #10 - 会員マイページとStripeポータルの実装
 */
export default async function MembershipPage() {
  // セッション情報を取得
  const session = await getServerSession(authOptions)

  if (!session?.user) {
    // middlewareで保護されているため、ここには到達しないはず
    return null
  }

  // 会員情報を取得
  const membership = await getUserMembership()
  const isPaidMember = membership?.membership === 'paid'

  // 支払いステータスの表示
  const paymentStatusDisplay = {
    active: {
      label: '正常',
      color: 'text-green-600 bg-green-50',
      icon: CheckCircle,
      description: 'お支払いは正常に処理されています',
    },
    past_due: {
      label: '支払い遅延',
      color: 'text-yellow-600 bg-yellow-50',
      icon: AlertCircle,
      description: 'お支払いの確認が必要です',
    },
    canceled: {
      label: 'キャンセル済み',
      color: 'text-gray-600 bg-gray-50',
      icon: XCircle,
      description: 'サブスクリプションはキャンセルされました',
    },
    unpaid: {
      label: '未払い',
      color: 'text-red-600 bg-red-50',
      icon: AlertCircle,
      description: 'お支払いが確認できません',
    },
  }

  const currentPaymentStatus = membership?.paymentStatus || 'active'
  const paymentInfo =
    paymentStatusDisplay[
      currentPaymentStatus as keyof typeof paymentStatusDisplay
    ] || paymentStatusDisplay.active

  return (
    <div className="min-h-screen bg-gray-50 py-8">
      <div className="mx-auto max-w-4xl px-4 sm:px-6 lg:px-8">
        {/* 戻るリンク */}
        <Link
          href="/media/mypage"
          className="mb-6 inline-flex items-center gap-2 text-sm text-gray-600 hover:text-gray-900"
        >
          <ArrowLeft className="h-4 w-4" />
          マイページに戻る
        </Link>

        {/* ページタイトル */}
        <h1 className="mb-8 text-2xl font-bold text-gray-900">
          会員情報・お支払い
        </h1>

        {/* 現在のプラン */}
        <div className="mb-8 rounded-lg bg-white p-6 shadow-sm">
          <h2 className="mb-4 text-lg font-semibold text-gray-900">
            現在のプラン
          </h2>

          {isPaidMember ? (
            <div className="space-y-4">
              <div className="flex items-center justify-between">
                <div className="flex items-center gap-3">
                  <Crown className="h-8 w-8 text-yellow-500" />
                  <div>
                    <p className="text-xl font-bold text-gray-900">
                      有料会員プラン
                    </p>
                    <p className="text-gray-600">月額 1,980円（税込）</p>
                  </div>
                </div>
                <span
                  className={`inline-flex items-center gap-1 rounded-full px-3 py-1 text-sm font-medium ${paymentInfo.color}`}
                >
                  <paymentInfo.icon className="h-4 w-4" />
                  {paymentInfo.label}
                </span>
              </div>

              <p className="text-sm text-gray-600">{paymentInfo.description}</p>

              {membership.membershipUpdatedAt && (
                <div className="flex items-center gap-2 text-sm text-gray-500">
                  <Calendar className="h-4 w-4" />
                  <span>
                    {format(
                      new Date(membership.membershipUpdatedAt),
                      'yyyy年MM月dd日',
                      { locale: ja }
                    )}
                    に更新
                  </span>
                </div>
              )}
            </div>
          ) : (
            <div className="space-y-4">
              <div className="flex items-center gap-3">
                <div className="rounded-full bg-gray-100 p-2">
                  <Crown className="h-6 w-6 text-gray-400" />
                </div>
                <div>
                  <p className="text-xl font-bold text-gray-900">無料会員</p>
                  <p className="text-gray-600">無料記事のみ閲覧可能</p>
                </div>
              </div>

              <div className="rounded-lg bg-yellow-50 p-4">
                <p className="mb-3 text-sm font-medium text-yellow-900">
                  有料会員になると以下の特典があります：
                </p>
                <ul className="space-y-1 text-sm text-yellow-800">
                  <li>• すべての有料記事が読み放題</li>
                  <li>• 専門家による深い分析・調査レポート</li>
                  <li>• 最新の暗号資産・ブロックチェーン情報</li>
                </ul>
                <Link href="/register" className="mt-4 block">
                  <Button className="w-full">
                    <Crown className="mr-2 h-4 w-4" />
                    有料会員になる（月額1,980円）
                  </Button>
                </Link>
              </div>
            </div>
          )}
        </div>

        {/* 契約情報管理（有料会員のみ） */}
        {isPaidMember && membership.stripeCustomerId && (
          <div className="mb-8 rounded-lg bg-white p-6 shadow-sm">
            <h2 className="mb-4 text-lg font-semibold text-gray-900">
              契約情報管理
            </h2>

            <p className="mb-4 text-gray-600">
              お支払い方法の変更、請求書の確認、サブスクリプションのキャンセルなどは、
              Stripeカスタマーポータルから行えます。
            </p>

            <form action="/api/stripe/portal" method="POST">
              <Button
                type="submit"
                variant="outline"
                className="w-full sm:w-auto"
              >
                <CreditCard className="mr-2 h-4 w-4" />
                契約情報を管理する
              </Button>
            </form>

            <p className="mt-4 text-sm text-gray-500">
              ※ Stripeの安全なポータルページに移動します
            </p>
          </div>
        )}

        {/* 注意事項 */}
        <div className="rounded-lg bg-gray-100 p-6">
          <h3 className="mb-3 font-semibold text-gray-900">ご注意事項</h3>
          <ul className="space-y-2 text-sm text-gray-600">
            <li>
              •
              サブスクリプションは自動更新されます。キャンセルはいつでも可能です。
            </li>
            <li>
              •
              キャンセル後も、現在の請求期間の終了日まで有料会員の特典をご利用いただけます。
            </li>
            <li>
              • お支払い情報の変更後、反映までに数分かかる場合があります。
            </li>
            <li>
              • ご不明な点がございましたら、
              <Link
                href="/media/mypage/support"
                className="text-blue-600 hover:text-blue-800"
              >
                サポート
              </Link>
              までお問い合わせください。
            </li>
          </ul>
        </div>
      </div>
    </div>
  )
}
