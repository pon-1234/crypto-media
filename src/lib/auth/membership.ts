import { getServerSession } from 'next-auth';
import { authOptions } from '@/lib/auth/authOptions';
import { adminDb } from '@/lib/firebase/admin';

/**
 * 会員情報の型定義
 */
export interface Membership {
  userId: string;
  email: string;
  membership: 'free' | 'paid';
  membershipUpdatedAt?: string;
  stripeCustomerId?: string;
  stripeSubscriptionId?: string;
  paymentStatus?: string;
}

/**
 * ユーザーの会員ステータスを取得するヘルパー関数
 * 
 * @doc この関数は、現在のセッションからユーザーIDを取得し、
 * Firestoreからユーザーの会員情報を取得します。
 * 
 * @returns ユーザーの会員情報。未認証の場合はnullを返します。
 * 
 * @related src/lib/auth/authOptions.ts - NextAuth設定
 * @related src/app/api/stripe/webhook/route.ts - 会員ステータス更新ロジック
 */
export async function getUserMembership(): Promise<Membership | null> {
  try {
    // 現在のセッションを取得
    const session = await getServerSession(authOptions);
    
    if (!session?.user?.id) {
      // 未認証ユーザー（ゲスト）
      return null;
    }

    // FirestoreからユーザーDocumentを取得
    const userDoc = await adminDb.collection('users').doc(session.user.id).get();
    
    if (!userDoc.exists) {
      // ユーザーDocumentが存在しない場合
      console.error(`User document not found for ID: ${session.user.id}`);
      return null;
    }

    const userData = userDoc.data();
    
    return {
      userId: session.user.id,
      email: userData?.email || session.user.email,
      membership: userData?.membership || 'free',
      membershipUpdatedAt: userData?.membershipUpdatedAt,
      stripeCustomerId: userData?.stripeCustomerId,
      stripeSubscriptionId: userData?.stripeSubscriptionId,
      paymentStatus: userData?.paymentStatus,
    };
  } catch (error) {
    console.error('Error fetching user membership:', error);
    return null;
  }
}

/**
 * ユーザーが有料会員かどうかを判定する関数
 * 
 * @doc この関数は、getUserMembershipを使用してユーザーの会員ステータスを確認し、
 * 有料会員かどうかを判定します。
 * 
 * @returns 有料会員の場合true、それ以外（無料会員・ゲスト）の場合false
 * 
 * @related getUserMembership - 会員情報取得関数
 */
export async function isPaidMember(): Promise<boolean> {
  const membership = await getUserMembership();
  return membership?.membership === 'paid';
}

/**
 * 会員ステータスに基づいたアクセス権限をチェックする関数
 * 
 * @doc この関数は、コンテンツの必要な会員レベルと
 * ユーザーの現在の会員ステータスを比較してアクセス可否を判定します。
 * 
 * @param requiredLevel - コンテンツに必要な会員レベル ('public' | 'paid')
 * @returns アクセス可能な場合true、不可の場合false
 * 
 * @related getUserMembership - 会員情報取得関数
 */
export async function hasAccess(requiredLevel: 'public' | 'paid'): Promise<boolean> {
  // publicコンテンツは誰でもアクセス可能
  if (requiredLevel === 'public') {
    return true;
  }

  // paidコンテンツは有料会員のみアクセス可能
  return await isPaidMember();
}

/**
 * 会員ステータスの型定義
 */
export type MembershipStatus = {
  userId: string;
  email: string | null;
  membership: 'free' | 'paid';
  membershipUpdatedAt?: string;
  stripeCustomerId?: string;
  stripeSubscriptionId?: string;
  paymentStatus?: 'active' | 'past_due' | 'canceled' | 'unpaid';
} | null;