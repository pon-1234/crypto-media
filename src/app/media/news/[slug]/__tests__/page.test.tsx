/**
 * @vitest-environment jsdom
 */
import { describe, it, expect, vi } from 'vitest'
import { notFound } from 'next/navigation'
import { draftMode } from 'next/headers'
import MediaNewsDetailPage, { generateMetadata } from '../page'
import { getMediaArticleBySlug, getRelatedArticles } from '@/lib/microcms'
import { hasAccess } from '@/lib/auth/membership'
import type { MediaArticle } from '@/lib/schema'
import DOMPurify from 'isomorphic-dompurify'

vi.mock('next/navigation', () => ({
  notFound: vi.fn(() => {
    throw new Error('NEXT_NOT_FOUND')
  }),
}))

vi.mock('next/headers', () => ({
  draftMode: vi.fn(() => ({
    isEnabled: false,
    enable: vi.fn(),
    disable: vi.fn(),
  })),
}))

vi.mock('@/lib/microcms', () => ({
  getMediaArticleBySlug: vi.fn(),
  getRelatedArticles: vi.fn(),
  getOptimizedImageUrl: vi.fn((url) => url),
}))

vi.mock('@/lib/auth/membership', () => ({
  hasAccess: vi.fn(),
}))

vi.mock('@/components/media/ArticleCard', () => ({
  ArticleCard: vi.fn(({ article }) => (
    <div data-testid={`related-${article.id}`}>{article.title}</div>
  )),
}))

vi.mock('@/components/media/Paywall', () => ({
  Paywall: vi.fn(({ article }) => (
    <div data-testid="paywall">Paywall for {article.title}</div>
  )),
}))

vi.mock('isomorphic-dompurify', () => ({
  default: {
    sanitize: vi.fn((html) => html),
  },
}))

describe('MediaNewsDetailPage', () => {
  const mockArticle: MediaArticle = {
    id: '1',
    title: 'メディアニュース記事',
    slug: 'media-news-article',
    type: 'media_news',
    membershipLevel: 'public',
    content: '<p>記事の内容</p>',
    heroImage: { url: 'https://example.com/image.jpg', width: 1200, height: 630 },
    category: { id: 'cat1', name: 'お知らせ', slug: 'announcement', createdAt: '2024-01-01T00:00:00Z', updatedAt: '2024-01-01T00:00:00Z' },
    tags: [{ id: 'tag1', name: 'Update', slug: 'update', createdAt: '2024-01-01T00:00:00Z', updatedAt: '2024-01-01T00:00:00Z' }],
    author: { id: 'author1', name: '執筆者', slug: 'author', role: ['執筆者'], profile: '', avatar: { url: '', width: 100, height: 100 }, createdAt: '2024-01-01T00:00:00Z', updatedAt: '2024-01-01T00:00:00Z' },
    createdAt: '2024-01-01T00:00:00Z',
    updatedAt: '2024-01-01T00:00:00Z',
    publishedAt: '2024-01-01T00:00:00Z',
    revisedAt: '2024-01-01T00:00:00Z',
  }

  const mockArticleWithSupervisor: MediaArticle = {
    ...mockArticle,
    supervisor: {
      id: 'supervisor1',
      name: '監修者',
      slug: 'supervisor',
      role: ['監修者'],
      profile: '',
      avatar: { url: '', width: 100, height: 100 },
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  }

  const mockRelatedArticles: MediaArticle[] = [
    { ...mockArticle, id: '2', title: '関連記事1', slug: 'related-1' },
    { ...mockArticle, id: '3', title: '関連記事2', slug: 'related-2' },
  ]

  it('メディアニュース記事を正しく表示する', async () => {
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(mockArticle)
    vi.mocked(getRelatedArticles).mockResolvedValue(mockRelatedArticles)
    vi.mocked(hasAccess).mockResolvedValue(true)

    const page = await MediaNewsDetailPage({
      params: { slug: 'media-news-article' },
      searchParams: {},
    })

    expect(getMediaArticleBySlug).toHaveBeenCalledWith('media-news-article', undefined)
    expect(page).toMatchSnapshot()
  })

  it('supervisorとtagsを持つ記事を正しく表示する', async () => {
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(mockArticleWithSupervisor)
    vi.mocked(getRelatedArticles).mockResolvedValue(mockRelatedArticles)
    vi.mocked(hasAccess).mockResolvedValue(true)

    const page = await MediaNewsDetailPage({
      params: { slug: 'media-news-article-with-supervisor' },
      searchParams: {},
    })

    expect(page).toMatchSnapshot()
  })

  it('関連記事の取得に失敗してもページが正常に表示される', async () => {
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(mockArticle)
    vi.mocked(getRelatedArticles).mockRejectedValue(new Error('API Error'))
    vi.mocked(hasAccess).mockResolvedValue(true)
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

    const page = await MediaNewsDetailPage({
      params: { slug: 'media-news-article' },
      searchParams: {},
    })

    expect(page).toMatchSnapshot()
    expect(consoleErrorSpy).toHaveBeenCalledWith(
      'Failed to fetch related articles:',
      expect.any(Error)
    )
    consoleErrorSpy.mockRestore()
  })

  it('type が media_news でない場合は notFound を呼ぶ', async () => {
    const nonNewsArticle = { ...mockArticle, type: 'article' as const }
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(nonNewsArticle)

    await expect(
      MediaNewsDetailPage({
        params: { slug: 'not-news' },
        searchParams: {},
      })
    ).rejects.toThrow('NEXT_NOT_FOUND')

    expect(notFound).toHaveBeenCalled()
  })

  it('記事が見つからない場合は notFound を呼ぶ', async () => {
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(null)

    await expect(
      MediaNewsDetailPage({
        params: { slug: 'not-found' },
        searchParams: {},
      })
    ).rejects.toThrow('NEXT_NOT_FOUND')

    expect(notFound).toHaveBeenCalled()
  })

  it('アクセス権限がない場合は Paywall を表示する', async () => {
    const paidArticle = { ...mockArticle, membershipLevel: 'paid' as const }
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(paidArticle)
    vi.mocked(getRelatedArticles).mockResolvedValue([])
    vi.mocked(hasAccess).mockResolvedValue(false)

    const page = await MediaNewsDetailPage({
      params: { slug: 'paid-article' },
      searchParams: {},
    })

    expect(hasAccess).toHaveBeenCalledWith('paid')
    expect(page).toMatchSnapshot()
  })

  it('プレビューモードで記事を表示する', async () => {
    vi.mocked(draftMode).mockReturnValue({
      isEnabled: true,
      enable: vi.fn(),
      disable: vi.fn(),
    })
    vi.mocked(getMediaArticleBySlug).mockResolvedValue(mockArticle)
    vi.mocked(getRelatedArticles).mockResolvedValue([])
    vi.mocked(hasAccess).mockResolvedValue(true)

    await MediaNewsDetailPage({
      params: { slug: 'preview-article' },
      searchParams: { draftKey: 'test-key' },
    })

    expect(getMediaArticleBySlug).toHaveBeenCalledWith('preview-article', { draftKey: 'test-key' })
  })

  describe('generateMetadata', () => {
    let originalCI: string | undefined

    beforeEach(() => {
      originalCI = process.env.CI
      process.env.CI = undefined
    })

    afterEach(() => {
      process.env.CI = originalCI
    })

    it('正しいメタデータを生成する', async () => {
      vi.mocked(getMediaArticleBySlug).mockResolvedValue(mockArticle)
      vi.mocked(DOMPurify.sanitize).mockReturnValue('記事の内容')

      const metadata = await generateMetadata({
        params: { slug: 'media-news-article' },
        searchParams: {},
      })

      expect(metadata).toEqual({
        title: 'メディアニュース記事 | Crypto Media',
        description: '記事の内容',
        openGraph: expect.objectContaining({
          title: 'メディアニュース記事',
          type: 'article',
          authors: ['執筆者'],
        }),
        twitter: expect.objectContaining({
          card: 'summary_large_image',
          title: 'メディアニュース記事',
        }),
      })
    })

    it('authorがいない記事のメタデータを正しく生成する', async () => {
      const articleWithoutAuthor = { ...mockArticle, author: undefined }
      vi.mocked(getMediaArticleBySlug).mockResolvedValue(articleWithoutAuthor)

      const metadata = await generateMetadata({
        params: { slug: 'media-news-article' },
        searchParams: {},
      })

      // @ts-expect-error: `authors` is a valid property but may not be in the current type definition
      expect(metadata.openGraph?.authors).toBeUndefined()
    })

    it('type が media_news でない場合はエラーメタデータを返す', async () => {
      const nonNewsArticle = { ...mockArticle, type: 'article' as const }
      vi.mocked(getMediaArticleBySlug).mockResolvedValue(nonNewsArticle)

      const metadata = await generateMetadata({
        params: { slug: 'not-news' },
        searchParams: {},
      })

      expect(metadata).toEqual({
        title: 'ニュースが見つかりません | Crypto Media',
      })
    })

    it('CI環境ではデフォルトメタデータを返す', async () => {
      process.env.CI = 'true'

      const metadata = await generateMetadata({
        params: { slug: 'test' },
        searchParams: {},
      })

      expect(metadata).toEqual({
        title: 'メディアニュース | Crypto Media',
        description: '暗号資産・ブロックチェーンに関するニュースの詳細をご覧いただけます。',
      })
    })
  })
})