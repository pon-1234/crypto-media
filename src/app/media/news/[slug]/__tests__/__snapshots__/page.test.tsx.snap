// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaNewsDetailPage > アクセス権限がない場合は Paywall を表示する 1`] = `
<article
  className="min-h-screen bg-white"
>
  <div
    className="relative aspect-[16/9] w-full"
  >
    <ForwardRef
      alt="メディアニュース記事"
      className="object-cover"
      fill={true}
      priority={true}
      sizes="100vw"
      src="https://example.com/image.jpg"
    />
  </div>
  <div
    className="mx-auto max-w-4xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <header
      className="mb-8"
    >
      <ForwardRef(LinkComponent)
        className="mb-2 inline-block text-sm font-medium text-blue-600 hover:text-blue-700"
        href="/media/category/announcement"
      >
        お知らせ
      </ForwardRef(LinkComponent)>
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        メディアニュース記事
      </h1>
      <div
        className="flex flex-wrap items-center gap-4 text-sm text-gray-600"
      >
        <time
          dateTime="2024-01-01T00:00:00Z"
        >
          2024年01月01日
        </time>
        <ForwardRef(LinkComponent)
          className="hover:text-gray-900"
          href="/media/experts/author"
        >
          執筆: 
          執筆者
        </ForwardRef(LinkComponent)>
        <span
          className="rounded bg-yellow-100 px-2 py-1 text-xs font-medium text-yellow-800"
        >
          有料会員限定
        </span>
      </div>
      <div
        className="mt-4 flex flex-wrap gap-2"
      >
        <ForwardRef(LinkComponent)
          className="rounded-full bg-gray-100 px-3 py-1 text-xs font-medium text-gray-700 hover:bg-gray-200"
          href="/media/tag/update"
        >
          #
          Update
        </ForwardRef(LinkComponent)>
      </div>
    </header>
    <spy
      preview="<p>記事の内容</p>..."
      title="メディアニュース記事"
    />
  </div>
</article>
`;

exports[`MediaNewsDetailPage > メディアニュース記事を正しく表示する 1`] = `
<article
  className="min-h-screen bg-white"
>
  <div
    className="relative aspect-[16/9] w-full"
  >
    <ForwardRef
      alt="メディアニュース記事"
      className="object-cover"
      fill={true}
      priority={true}
      sizes="100vw"
      src="https://example.com/image.jpg"
    />
  </div>
  <div
    className="mx-auto max-w-4xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <header
      className="mb-8"
    >
      <ForwardRef(LinkComponent)
        className="mb-2 inline-block text-sm font-medium text-blue-600 hover:text-blue-700"
        href="/media/category/announcement"
      >
        お知らせ
      </ForwardRef(LinkComponent)>
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        メディアニュース記事
      </h1>
      <div
        className="flex flex-wrap items-center gap-4 text-sm text-gray-600"
      >
        <time
          dateTime="2024-01-01T00:00:00Z"
        >
          2024年01月01日
        </time>
        <ForwardRef(LinkComponent)
          className="hover:text-gray-900"
          href="/media/experts/author"
        >
          執筆: 
          執筆者
        </ForwardRef(LinkComponent)>
      </div>
      <div
        className="mt-4 flex flex-wrap gap-2"
      >
        <ForwardRef(LinkComponent)
          className="rounded-full bg-gray-100 px-3 py-1 text-xs font-medium text-gray-700 hover:bg-gray-200"
          href="/media/tag/update"
        >
          #
          Update
        </ForwardRef(LinkComponent)>
      </div>
    </header>
    <div
      className="prose prose-lg max-w-none"
      dangerouslySetInnerHTML={
        {
          "__html": "<p>記事の内容</p>",
        }
      }
    />
    <section
      className="mt-12 border-t pt-8"
    >
      <h2
        className="mb-6 text-xl font-bold text-gray-900 sm:text-2xl"
      >
        関連記事
      </h2>
      <div
        className="grid grid-cols-1 gap-6 md:grid-cols-3"
      >
        <spy
          article={
            {
              "author": {
                "avatar": {
                  "height": 100,
                  "url": "",
                  "width": 100,
                },
                "createdAt": "2024-01-01T00:00:00Z",
                "id": "author1",
                "name": "執筆者",
                "profile": "",
                "role": [
                  "執筆者",
                ],
                "slug": "author",
                "updatedAt": "2024-01-01T00:00:00Z",
              },
              "category": {
                "createdAt": "2024-01-01T00:00:00Z",
                "id": "cat1",
                "name": "お知らせ",
                "slug": "announcement",
                "updatedAt": "2024-01-01T00:00:00Z",
              },
              "content": "<p>記事の内容</p>",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image.jpg",
                "width": 1200,
              },
              "id": "2",
              "membershipLevel": "public",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "related-1",
              "tags": [
                {
                  "createdAt": "2024-01-01T00:00:00Z",
                  "id": "tag1",
                  "name": "Update",
                  "slug": "update",
                  "updatedAt": "2024-01-01T00:00:00Z",
                },
              ],
              "title": "関連記事1",
              "type": "media_news",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
        <spy
          article={
            {
              "author": {
                "avatar": {
                  "height": 100,
                  "url": "",
                  "width": 100,
                },
                "createdAt": "2024-01-01T00:00:00Z",
                "id": "author1",
                "name": "執筆者",
                "profile": "",
                "role": [
                  "執筆者",
                ],
                "slug": "author",
                "updatedAt": "2024-01-01T00:00:00Z",
              },
              "category": {
                "createdAt": "2024-01-01T00:00:00Z",
                "id": "cat1",
                "name": "お知らせ",
                "slug": "announcement",
                "updatedAt": "2024-01-01T00:00:00Z",
              },
              "content": "<p>記事の内容</p>",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image.jpg",
                "width": 1200,
              },
              "id": "3",
              "membershipLevel": "public",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "related-2",
              "tags": [
                {
                  "createdAt": "2024-01-01T00:00:00Z",
                  "id": "tag1",
                  "name": "Update",
                  "slug": "update",
                  "updatedAt": "2024-01-01T00:00:00Z",
                },
              ],
              "title": "関連記事2",
              "type": "media_news",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
      </div>
    </section>
  </div>
</article>
`;
